File: CPlus/read4II.cpp
int read4(char *buf);

class Solution {
public:
    /**
     * @param buf Destination buffer
     * @param n   Number of characters to read
     * @return    The number of actual characters read
     */
    int read(char *buf, int n) {
        for (int i=0;i<n;++i){
            if(read_pos==write_pos){
                write_pos=read4(buff);
                read_pos=0;
                if(write_pos==0) return i;
            }
            
            buf[i]=buff[read_pos++];
        }
        return n;
    }
private:
    int read_pos=0,write_pos=0;
    char buff[4];
};

File: Python/read4II.py
class Solution(object):
    def __init__(self):
        self.buf=['']*4
        self.index=0
        self.buf_size=0
    
    def read(self, buf, n):
        """
        :type buf: Destination buffer (List[str])
        :type n: Number of characters to read (int)
        :rtype: The number of actual characters read (int)
        """
        index=0
        
        while index<n:
            while index<n and self.index<self.buf_size:
                buf[index]=self.buf[self.index]
                index+=1
                self.index+=1
            
            if self.index==self.buf_size:
                self.buf_size=read4(self.buf)
                self.index = 0
            
            if self.buf_size == 0:
                break
        
        return index